src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs

```diff
@@ -0,0 +1,236 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.EngineV1
+{
+    internal partial class DiagnosticIncrementalAnalyzer
+    {
+        private static readonly int s_stateTypeCount = Enum.GetNames(typeof(StateType)).Count();
+
+        /// <summary>
+        /// This is in charge of anything related to <see cref="DiagnosticState"/>
+        /// </summary>
+        private partial class StateManager
+        {
+            private readonly HostAnalyzerManager _analyzerManager;
+
+            private readonly HostStates _hostStates;
+            private readonly ProjectStates _projectStates;
+
+            public StateManager(HostAnalyzerManager analyzerManager)
+            {
+                _analyzerManager = analyzerManager;
+
+                _hostStates = new HostStates(this);
+                _projectStates = new ProjectStates(this);
+            }
+
+            private HostAnalyzerManager AnalyzerManager { get { return _analyzerManager; } }
+
+            /// <summary>
+            /// This will be raised whenever <see cref="StateManager"/> finds <see cref="Project.AnalyzerReferences"/> change
+            /// </summary>
+            public event EventHandler<ProjectAnalyzerReferenceChangedEventArgs> ProjectAnalyzerReferenceChanged;
+
+            /// <summary>
+            /// Return <see cref="StateSet"/>s for the given <see cref="ProjectId"/>. 
+            /// This will never create new <see cref="StateSet"/> but will return ones already created.
+            /// </summary>
+            public IEnumerable<StateSet> GetStateSets(ProjectId projectId)
+            {
+                return _hostStates.GetStateSets().Concat(_projectStates.GetStateSets(projectId));
+            }
+
+            /// <summary>
+            /// Return <see cref="StateSet"/>s for the given <see cref="Project"/>.
+            /// This will never create new <see cref="StateSet"/> but will return ones already created.
+            /// Difference with <see cref="GetStateSets(ProjectId)"/> is that 
+            /// this will only return <see cref="StateSet"/>s that have same language as <paramref name="project"/>.
+            /// </summary>
+            public IEnumerable<StateSet> GetStateSets(Project project)
```

Comment:
I think pattern of Get and GetOrCreate seems clearer. otherwise I think it should be GetExistingXXX and GetXXX